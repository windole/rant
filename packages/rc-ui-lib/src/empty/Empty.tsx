import React, { isValidElement, useContext } from 'react';
import classnames from 'classnames';
import { EmptyProps } from './PropsType';
import { getSizeStyle } from '../utils';
import { Network } from './Network';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';

const PRESET_IMAGES = ['error', 'search', 'default'];

const Empty: React.FC<EmptyProps> = (props) => {
  const { prefixCls, createNamespace } = useContext(ConfigProviderContext);
  const [bem] = createNamespace('empty', prefixCls);

  const renderImage = () => {
    let { image } = props;

    if (isValidElement(image)) {
      return image;
    }

    if (image === 'network') {
      return Network;
    }

    if (PRESET_IMAGES.includes(image as string)) {
      image = `https://img.yzcdn.cn/vant/empty-image-${image}.png`;
    }

    return <img src={image as string} alt="" />;
  };

  const renderDescription = () => {
    if (props.description) {
      return <p className={classnames(bem('description'))}>{props.description}</p>;
    }
    return null;
  };

  const renderBottom = () => {
    if (props.children) {
      return <div className={classnames(bem('bottom'))}>{props.children}</div>;
    }
    return null;
  };

  return (
    <div className={classnames(props.className, bem())} style={props.style}>
      <div className={classnames(bem('image'))} style={getSizeStyle(props.imageSize)}>
        {renderImage()}
      </div>
      {renderDescription()}
      {renderBottom()}
    </div>
  );
};

Empty.defaultProps = {
  image: 'default',
};

export default Empty;
