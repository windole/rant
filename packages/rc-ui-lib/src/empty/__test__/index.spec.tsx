import * as React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Button from '../../button';
import { Network } from '../Network';
import Empty from '..';

describe('Empty', () => {
  let wrapper;
  afterEach(() => {
    wrapper.unmount();
  });

  it('should render image correctly', () => {
    wrapper = mount(<Empty image="search" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render description correctly', () => {
    wrapper = mount(<Empty description="description" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render bottom correctly', () => {
    wrapper = mount(
      <Empty>
        <Button round type="primary" className="bottom-button">
          按钮
        </Button>
      </Empty>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render svg when image is network component', () => {
    wrapper = mount(<Empty image={Network} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render svg when prop image is network', () => {
    wrapper = mount(<Empty image="network" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should change image size when using image-size prop', async () => {
    wrapper = mount(<Empty image="search" imageSize={50} />);

    const image = wrapper.find('.rc-empty__image');

    expect(image.getDOMNode().style.width).toEqual('50px');
    expect(image.getDOMNode().style.height).toEqual('50px');

    await wrapper.setProps({ imageSize: '1vw' });
    expect(image.getDOMNode().style.width).toEqual('1vw');
    expect(image.getDOMNode().style.height).toEqual('1vw');
  });
});
