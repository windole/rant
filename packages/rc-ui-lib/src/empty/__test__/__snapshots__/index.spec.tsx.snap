// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Empty should render bottom correctly 1`] = `
<Empty
  image="default"
>
  <div
    className="rc-empty"
  >
    <div
      className="rc-empty__image"
      style={Object {}}
    >
      <img
        alt=""
        src="https://img.yzcdn.cn/vant/empty-image-default.png"
      />
    </div>
    <div
      className="rc-empty__bottom"
    >
      <Button
        className="bottom-button"
        iconPosition="left"
        round={true}
        size="normal"
        tag="button"
        type="primary"
      >
        <button
          className="bottom-button rc-button rc-button--primary rc-button--normal rc-button--round"
          onClick={[Function]}
          style={Object {}}
        >
          <div
            className="rc-button__content"
          >
            <span
              className="rc-button__text"
              key="text"
            >
              按钮
            </span>
          </div>
        </button>
      </Button>
    </div>
  </div>
</Empty>
`;

exports[`Empty should render description correctly 1`] = `
<Empty
  description="description"
  image="default"
>
  <div
    className="rc-empty"
  >
    <div
      className="rc-empty__image"
      style={Object {}}
    >
      <img
        alt=""
        src="https://img.yzcdn.cn/vant/empty-image-default.png"
      />
    </div>
    <p
      className="rc-empty__description"
    >
      description
    </p>
  </div>
</Empty>
`;

exports[`Empty should render image correctly 1`] = `
<Empty
  image="search"
>
  <div
    className="rc-empty"
  >
    <div
      className="rc-empty__image"
      style={Object {}}
    >
      <img
        alt=""
        src="https://img.yzcdn.cn/vant/empty-image-search.png"
      />
    </div>
  </div>
</Empty>
`;

exports[`Empty should render svg when image is network component 1`] = `
<Empty
  image={
    <svg
      viewBox="0 0 160 160"
    >
      <defs>
        <linearGradient
          id="rc-empty-network-1"
          x1="64.022%"
          x2="64.022%"
          y1="100%"
          y2="0%"
        >
          <stop
            offset="0%"
            stopColor="#FFF"
            stopOpacity={0.5}
          />
          <stop
            offset="100%"
            stopColor="#F2F3F5"
          />
        </linearGradient>
        <linearGradient
          id="rc-empty-network-2"
          x1="50%"
          x2="50%"
          y1="0%"
          y2="84.459%"
        >
          <stop
            offset="0%"
            stopColor="#EBEDF0"
          />
          <stop
            offset="100%"
            stopColor="#DCDEE0"
            stopOpacity={0}
          />
        </linearGradient>
        <linearGradient
          id="rc-empty-network-3"
          x1="100%"
          x2="100%"
          y1="0%"
          y2="100%"
        >
          <stop
            offset="0%"
            stopColor="#EAEDF0"
          />
          <stop
            offset="100%"
            stopColor="#DCDEE0"
          />
        </linearGradient>
        <linearGradient
          id="rc-empty-network-4"
          x1="100%"
          x2="100%"
          y1="100%"
          y2="0%"
        >
          <stop
            offset="0%"
            stopColor="#EAEDF0"
          />
          <stop
            offset="100%"
            stopColor="#DCDEE0"
          />
        </linearGradient>
        <linearGradient
          id="rc-empty-network-5"
          x1="0%"
          x2="100%"
          y1="43.982%"
          y2="54.703%"
        >
          <stop
            offset="0%"
            stopColor="#EAEDF0"
          />
          <stop
            offset="100%"
            stopColor="#DCDEE0"
          />
        </linearGradient>
        <linearGradient
          id="rc-empty-network-6"
          x1="94.535%"
          x2="5.465%"
          y1="43.837%"
          y2="54.948%"
        >
          <stop
            offset="0%"
            stopColor="#EAEDF0"
          />
          <stop
            offset="100%"
            stopColor="#DCDEE0"
          />
        </linearGradient>
        <radialGradient
          cx="50%"
          cy="0%"
          fx="50%"
          fy="0%"
          gradientTransform="matrix(0 1 -.54835 0 .5 -.5)"
          id="rc-empty-network-7"
          r="100%"
        >
          <stop
            offset="0%"
            stopColor="#EBEDF0"
          />
          <stop
            offset="100%"
            stopColor="#FFF"
            stopOpacity={0}
          />
        </radialGradient>
      </defs>
      <g
        fill="none"
        fillRule="evenodd"
      >
        <g
          opacity=".8"
        >
          <path
            d="M0 124V46h20v20h14v58H0z"
            fill="url(#rc-empty-network-1)"
            transform="matrix(-1 0 0 1 36 7)"
          />
          <path
            d="M121 8h22.231v14H152v77.37h-31V8z"
            fill="url(#rc-empty-network-1)"
            transform="translate(2 7)"
          />
        </g>
        <path
          d="M0 139h160v21H0z"
          fill="url(#rc-empty-network-7)"
        />
        <path
          d="M37 18a7 7 0 013 13.326v26.742c0 1.23-.997 2.227-2.227 2.227h-1.546A2.227 2.227 0 0134 58.068V31.326A7 7 0 0137 18z"
          fill="url(#rc-empty-network-2)"
          fillRule="nonzero"
          transform="translate(43 36)"
        />
        <g
          opacity=".6"
          strokeLinecap="round"
          strokeWidth="7"
        >
          <path
            d="M20.875 11.136a18.868 18.868 0 00-5.284 13.121c0 5.094 2.012 9.718 5.284 13.12"
            stroke="url(#rc-empty-network-3)"
            transform="translate(43 36)"
          />
          <path
            d="M9.849 0C3.756 6.225 0 14.747 0 24.146c0 9.398 3.756 17.92 9.849 24.145"
            stroke="url(#rc-empty-network-3)"
            transform="translate(43 36)"
          />
          <path
            d="M57.625 11.136a18.868 18.868 0 00-5.284 13.121c0 5.094 2.012 9.718 5.284 13.12"
            stroke="url(#rc-empty-network-4)"
            transform="rotate(-180 76.483 42.257)"
          />
          <path
            d="M73.216 0c-6.093 6.225-9.849 14.747-9.849 24.146 0 9.398 3.756 17.92 9.849 24.145"
            stroke="url(#rc-empty-network-4)"
            transform="rotate(-180 89.791 42.146)"
          />
        </g>
        <g
          fillRule="nonzero"
          transform="translate(31 105)"
        >
          <rect
            fill="url(#rc-empty-network-5)"
            height="34"
            rx="2"
            width="98"
          />
          <rect
            fill="#FFF"
            height="18"
            rx="1.114"
            width="80"
            x="9"
            y="8"
          />
          <rect
            fill="url(#rc-empty-network-6)"
            height="6"
            rx="1.114"
            width="18"
            x="15"
            y="12"
          />
        </g>
      </g>
    </svg>
  }
>
  <div
    className="rc-empty"
  >
    <div
      className="rc-empty__image"
      style={Object {}}
    >
      <svg
        viewBox="0 0 160 160"
      >
        <defs>
          <linearGradient
            id="rc-empty-network-1"
            x1="64.022%"
            x2="64.022%"
            y1="100%"
            y2="0%"
          >
            <stop
              offset="0%"
              stopColor="#FFF"
              stopOpacity={0.5}
            />
            <stop
              offset="100%"
              stopColor="#F2F3F5"
            />
          </linearGradient>
          <linearGradient
            id="rc-empty-network-2"
            x1="50%"
            x2="50%"
            y1="0%"
            y2="84.459%"
          >
            <stop
              offset="0%"
              stopColor="#EBEDF0"
            />
            <stop
              offset="100%"
              stopColor="#DCDEE0"
              stopOpacity={0}
            />
          </linearGradient>
          <linearGradient
            id="rc-empty-network-3"
            x1="100%"
            x2="100%"
            y1="0%"
            y2="100%"
          >
            <stop
              offset="0%"
              stopColor="#EAEDF0"
            />
            <stop
              offset="100%"
              stopColor="#DCDEE0"
            />
          </linearGradient>
          <linearGradient
            id="rc-empty-network-4"
            x1="100%"
            x2="100%"
            y1="100%"
            y2="0%"
          >
            <stop
              offset="0%"
              stopColor="#EAEDF0"
            />
            <stop
              offset="100%"
              stopColor="#DCDEE0"
            />
          </linearGradient>
          <linearGradient
            id="rc-empty-network-5"
            x1="0%"
            x2="100%"
            y1="43.982%"
            y2="54.703%"
          >
            <stop
              offset="0%"
              stopColor="#EAEDF0"
            />
            <stop
              offset="100%"
              stopColor="#DCDEE0"
            />
          </linearGradient>
          <linearGradient
            id="rc-empty-network-6"
            x1="94.535%"
            x2="5.465%"
            y1="43.837%"
            y2="54.948%"
          >
            <stop
              offset="0%"
              stopColor="#EAEDF0"
            />
            <stop
              offset="100%"
              stopColor="#DCDEE0"
            />
          </linearGradient>
          <radialGradient
            cx="50%"
            cy="0%"
            fx="50%"
            fy="0%"
            gradientTransform="matrix(0 1 -.54835 0 .5 -.5)"
            id="rc-empty-network-7"
            r="100%"
          >
            <stop
              offset="0%"
              stopColor="#EBEDF0"
            />
            <stop
              offset="100%"
              stopColor="#FFF"
              stopOpacity={0}
            />
          </radialGradient>
        </defs>
        <g
          fill="none"
          fillRule="evenodd"
        >
          <g
            opacity=".8"
          >
            <path
              d="M0 124V46h20v20h14v58H0z"
              fill="url(#rc-empty-network-1)"
              transform="matrix(-1 0 0 1 36 7)"
            />
            <path
              d="M121 8h22.231v14H152v77.37h-31V8z"
              fill="url(#rc-empty-network-1)"
              transform="translate(2 7)"
            />
          </g>
          <path
            d="M0 139h160v21H0z"
            fill="url(#rc-empty-network-7)"
          />
          <path
            d="M37 18a7 7 0 013 13.326v26.742c0 1.23-.997 2.227-2.227 2.227h-1.546A2.227 2.227 0 0134 58.068V31.326A7 7 0 0137 18z"
            fill="url(#rc-empty-network-2)"
            fillRule="nonzero"
            transform="translate(43 36)"
          />
          <g
            opacity=".6"
            strokeLinecap="round"
            strokeWidth="7"
          >
            <path
              d="M20.875 11.136a18.868 18.868 0 00-5.284 13.121c0 5.094 2.012 9.718 5.284 13.12"
              stroke="url(#rc-empty-network-3)"
              transform="translate(43 36)"
            />
            <path
              d="M9.849 0C3.756 6.225 0 14.747 0 24.146c0 9.398 3.756 17.92 9.849 24.145"
              stroke="url(#rc-empty-network-3)"
              transform="translate(43 36)"
            />
            <path
              d="M57.625 11.136a18.868 18.868 0 00-5.284 13.121c0 5.094 2.012 9.718 5.284 13.12"
              stroke="url(#rc-empty-network-4)"
              transform="rotate(-180 76.483 42.257)"
            />
            <path
              d="M73.216 0c-6.093 6.225-9.849 14.747-9.849 24.146 0 9.398 3.756 17.92 9.849 24.145"
              stroke="url(#rc-empty-network-4)"
              transform="rotate(-180 89.791 42.146)"
            />
          </g>
          <g
            fillRule="nonzero"
            transform="translate(31 105)"
          >
            <rect
              fill="url(#rc-empty-network-5)"
              height="34"
              rx="2"
              width="98"
            />
            <rect
              fill="#FFF"
              height="18"
              rx="1.114"
              width="80"
              x="9"
              y="8"
            />
            <rect
              fill="url(#rc-empty-network-6)"
              height="6"
              rx="1.114"
              width="18"
              x="15"
              y="12"
            />
          </g>
        </g>
      </svg>
    </div>
  </div>
</Empty>
`;

exports[`Empty should render svg when prop image is network 1`] = `
<Empty
  image="network"
>
  <div
    className="rc-empty"
  >
    <div
      className="rc-empty__image"
      style={Object {}}
    >
      <svg
        viewBox="0 0 160 160"
      >
        <defs>
          <linearGradient
            id="rc-empty-network-1"
            x1="64.022%"
            x2="64.022%"
            y1="100%"
            y2="0%"
          >
            <stop
              offset="0%"
              stopColor="#FFF"
              stopOpacity={0.5}
            />
            <stop
              offset="100%"
              stopColor="#F2F3F5"
            />
          </linearGradient>
          <linearGradient
            id="rc-empty-network-2"
            x1="50%"
            x2="50%"
            y1="0%"
            y2="84.459%"
          >
            <stop
              offset="0%"
              stopColor="#EBEDF0"
            />
            <stop
              offset="100%"
              stopColor="#DCDEE0"
              stopOpacity={0}
            />
          </linearGradient>
          <linearGradient
            id="rc-empty-network-3"
            x1="100%"
            x2="100%"
            y1="0%"
            y2="100%"
          >
            <stop
              offset="0%"
              stopColor="#EAEDF0"
            />
            <stop
              offset="100%"
              stopColor="#DCDEE0"
            />
          </linearGradient>
          <linearGradient
            id="rc-empty-network-4"
            x1="100%"
            x2="100%"
            y1="100%"
            y2="0%"
          >
            <stop
              offset="0%"
              stopColor="#EAEDF0"
            />
            <stop
              offset="100%"
              stopColor="#DCDEE0"
            />
          </linearGradient>
          <linearGradient
            id="rc-empty-network-5"
            x1="0%"
            x2="100%"
            y1="43.982%"
            y2="54.703%"
          >
            <stop
              offset="0%"
              stopColor="#EAEDF0"
            />
            <stop
              offset="100%"
              stopColor="#DCDEE0"
            />
          </linearGradient>
          <linearGradient
            id="rc-empty-network-6"
            x1="94.535%"
            x2="5.465%"
            y1="43.837%"
            y2="54.948%"
          >
            <stop
              offset="0%"
              stopColor="#EAEDF0"
            />
            <stop
              offset="100%"
              stopColor="#DCDEE0"
            />
          </linearGradient>
          <radialGradient
            cx="50%"
            cy="0%"
            fx="50%"
            fy="0%"
            gradientTransform="matrix(0 1 -.54835 0 .5 -.5)"
            id="rc-empty-network-7"
            r="100%"
          >
            <stop
              offset="0%"
              stopColor="#EBEDF0"
            />
            <stop
              offset="100%"
              stopColor="#FFF"
              stopOpacity={0}
            />
          </radialGradient>
        </defs>
        <g
          fill="none"
          fillRule="evenodd"
        >
          <g
            opacity=".8"
          >
            <path
              d="M0 124V46h20v20h14v58H0z"
              fill="url(#rc-empty-network-1)"
              transform="matrix(-1 0 0 1 36 7)"
            />
            <path
              d="M121 8h22.231v14H152v77.37h-31V8z"
              fill="url(#rc-empty-network-1)"
              transform="translate(2 7)"
            />
          </g>
          <path
            d="M0 139h160v21H0z"
            fill="url(#rc-empty-network-7)"
          />
          <path
            d="M37 18a7 7 0 013 13.326v26.742c0 1.23-.997 2.227-2.227 2.227h-1.546A2.227 2.227 0 0134 58.068V31.326A7 7 0 0137 18z"
            fill="url(#rc-empty-network-2)"
            fillRule="nonzero"
            transform="translate(43 36)"
          />
          <g
            opacity=".6"
            strokeLinecap="round"
            strokeWidth="7"
          >
            <path
              d="M20.875 11.136a18.868 18.868 0 00-5.284 13.121c0 5.094 2.012 9.718 5.284 13.12"
              stroke="url(#rc-empty-network-3)"
              transform="translate(43 36)"
            />
            <path
              d="M9.849 0C3.756 6.225 0 14.747 0 24.146c0 9.398 3.756 17.92 9.849 24.145"
              stroke="url(#rc-empty-network-3)"
              transform="translate(43 36)"
            />
            <path
              d="M57.625 11.136a18.868 18.868 0 00-5.284 13.121c0 5.094 2.012 9.718 5.284 13.12"
              stroke="url(#rc-empty-network-4)"
              transform="rotate(-180 76.483 42.257)"
            />
            <path
              d="M73.216 0c-6.093 6.225-9.849 14.747-9.849 24.146 0 9.398 3.756 17.92 9.849 24.145"
              stroke="url(#rc-empty-network-4)"
              transform="rotate(-180 89.791 42.146)"
            />
          </g>
          <g
            fillRule="nonzero"
            transform="translate(31 105)"
          >
            <rect
              fill="url(#rc-empty-network-5)"
              height="34"
              rx="2"
              width="98"
            />
            <rect
              fill="#FFF"
              height="18"
              rx="1.114"
              width="80"
              x="9"
              y="8"
            />
            <rect
              fill="url(#rc-empty-network-6)"
              height="6"
              rx="1.114"
              width="18"
              x="15"
              y="12"
            />
          </g>
        </g>
      </svg>
    </div>
  </div>
</Empty>
`;
